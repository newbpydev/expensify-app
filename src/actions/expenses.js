import { v4 as uuidv4 } from "uuid";
import database from "../firebase/firebase";
import {
  getDatabase,
  ref,
  set,
  update,
  remove,
  onValue,
  child,
  get,
  off,
  push,
  onChildRemoved,
  onChildChanged,
  onChildAdded,
} from "firebase/database";

//! //////////////////////////////////////////////////////////////////////////
//! ACTIONS

//* ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

// export const addExpense = ({
//   description = "",
//   note = "",
//   amount = 0,
//   createdAt = 0,
// } = {}) => ({
//   type: "ADD_EXPENSE",
//   expense: {
//     id: uuidv4(),
//     description,
//     note,
//     amount,
//     createdAt,
//   },
// });

/* */
//*  adding Expenses to database and Redux
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt };

    return push(ref(database, "expenses"), expense).then((ref) => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        })
      );
    });
  };
};

//* REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = (expenseId) => {
  return (dispatch) => {
    const id = expenseId

    return remove(ref(database, `expenses/${id}`))
      .then(() => {
        dispatch(removeExpense(id))
        console.log(`removed item`);
      })
      .catch((e) => console.log('delete failed. ', e))

  }
}

//* EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

//*SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch) => {
    return get(ref(database, "expenses"))
      .then((snapshot) => {
        const expenses = [];

        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });


        dispatch(setExpenses(expenses));
      })
  };
};
