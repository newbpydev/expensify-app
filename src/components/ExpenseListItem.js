import React from "react";
// import { connect } from "react-redux";

import { Link } from "react-router-dom";

const ExpenseListItem = ({id, description, amount, createdAt, expenses, dispatch }) => (
  <div>
    <h3><Link to={`edit/${id}`} state={{id}}>{description}</Link></h3>
    <p>
      Amount: {amount} - Created At: {createdAt}
    </p>
    
  </div>
);

// const mapStateToProps = (state) => {
//   return {
//     expenses: state.expenses
//   }
// }

export default ExpenseListItem;
