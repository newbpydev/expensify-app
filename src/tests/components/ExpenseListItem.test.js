import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import ExpenseListItem from "../../components/ExpenseListItem";

test("should render ExpenseListItem with expense", () => {
  const expense = expenses[1]
  const wrapper = shallow(<ExpenseListItem {...expense} />)
  // console.log(expense)
  expect(wrapper).toMatchSnapshot();
})








