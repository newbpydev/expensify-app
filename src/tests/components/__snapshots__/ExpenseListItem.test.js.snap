// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense 1`] = `
<div>
  <h3>
    <Link
      state={
        Object {
          "id": "2",
        }
      }
      to="/edit/2"
    >
      gas
    </Link>
  </h3>
  <p>
    Amount: 
    $19.00
    - Created At: 
    December 27th, 1969
  </p>
</div>
`;
